File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.java
import java.util.*;
public class Main {
    public static int n;
    public static int[][] arr;
    public static boolean[] selected;
    public static int ans = Integer.MAX_VALUE;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        arr = new int[n][n];
        selected = new boolean[n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                arr[i][j]=sc.nextInt();
            }
        }
        // 조합으로 모든 팀 탐색
        combination(0, 0);
        System.out.println(ans);
    }

    private static void combination(int cnt, int start) {
        if(cnt==n/2) {
            // 조합된 팀으로 능력치 계산
            calculate();
            return;
        }
        for (int i = start; i < n; i++) {
            selected[i]=true;
            combination(cnt+1, i+1);
            selected[i]=false;
        }
    }

    private static void calculate() {
        int startTeam=0, linkTeam=0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if(selected[i] && selected[j]) {
                    startTeam+=arr[i][j];
                } else if (!selected[i]&&!selected[j]) {
                    linkTeam+=arr[i][j];
                }
            }
        }
        ans=Math.min(Math.abs(startTeam-linkTeam), ans);
    }
}

File: 백준/Silver/14889. 스타트와 링크/스타트와 링크.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(filePath).toString().trim().split(/\s+/);

let idx = 0;
const n = Number(input[idx++]);
const arr = [];
for(let i=0; i<n; i++) {
    const row = [];
    for(let j=0; j<n; j++) {
        row.push(Number(input[idx++]));
    }
    arr.push(row);
}

let ans = Infinity;
const selected = [];

dfs(0, 0);

function dfs(count, start) {
    if(count==n/2) {
        const startTeam = [];
        const linkTeam = [];
        for(let i=0; i<n; i++) {
            if(!selected.includes(i)) {
                linkTeam.push(i); // 링크 팀의 조합
            } else {
                startTeam.push(i);  // 스타트 팀의 조합
            }
        }
        teamPower(startTeam, linkTeam);
        return;
    }
    
    for(let i=start; i<n; i++) {
        selected.push(i);
        dfs(count+1, i+1);
        selected.pop();
    }
}

function teamPower(startTeam, linkTeam) {
    let sumStart = 0;
    let sumLink = 0;
    
    for(let i=0; i<n/2; i++) {
        for(let j=0; j<n/2; j++) {
            if(i===j) continue;
            sumStart += arr[startTeam[i]][startTeam[j]];
            sumLink += arr[linkTeam[i]][linkTeam[j]];
        }
    }
    const diff = Math.abs(sumStart-sumLink);
    ans = Math.min(diff, ans);
}

console.log(ans);